realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp

```diff
@@ -501,3 +506,57 @@ JNIEXPORT void JNICALL Java_io_realm_internal_SharedRealm_nativeRegisterSchemaCh
         java_binding_context.set_schema_changed_callback(env, j_schema_changed_callback);
     }
 }
+
+JNIEXPORT void JNICALL Java_io_realm_internal_SharedRealm_nativeRegisterPartialSyncQuery(
+    REALM_UNUSED JNIEnv* env, jclass, REALM_UNUSED jlong shared_realm_ptr, REALM_UNUSED jstring j_class_name,
+    REALM_UNUSED jstring j_table_name, REALM_UNUSED jclass j_clazz, REALM_UNUSED jstring j_query,
+    REALM_UNUSED jobject j_callback, REALM_UNUSED jobject j_realm)
+{
+    TR_ENTER_PTR(shared_realm_ptr)
+
+#if REALM_ENABLE_SYNC
+
+    auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    try {
+        JStringAccessor class_name(env, j_class_name);               // throws
+        std::string table_name = JStringAccessor(env, j_table_name); // throws
+        JStringAccessor query(env, j_query);                         // throws
+
+        JavaGlobalWeakRef j_callback_weak(env, j_callback);
+        JavaGlobalWeakRef j_realm_weak(env, j_realm);
+        JavaGlobalWeakRef j_clazz_weak(env, j_clazz);
+        static JavaClass shared_realm_class(env, "io/realm/internal/SharedRealm");
+        static JavaMethod partial_sync_cb(env, shared_realm_class, "runPartialSyncRegistrationCallback",
+                                          "(Ljava/lang/String;JLjava/lang/String;Ljava/lang/Class;Lio/realm/"
+                                          "Realm$PartialSyncCallback;Lio/realm/Realm;)V",
+                                          true);
+
+        auto cb = [j_callback_weak, j_realm_weak, j_class_name, j_clazz_weak,
+                   table_name](REALM_UNUSED Results results, std::exception_ptr err) {
+            JNIEnv* env = JniUtils::get_env(true);
+            if (err) {
+                try {
+                    std::rethrow_exception(err);
+                }
+                catch (const std::exception& e) {
+                    env->CallStaticVoidMethod(shared_realm_class, partial_sync_cb, to_jstring(env, e.what()),
+                                              reinterpret_cast<jlong>(nullptr), nullptr, nullptr,
+                                              j_callback_weak.global_ref(env).get(), nullptr);
+                }
+                return;
+            }
+
+            auto wrapper = new ResultsWrapper(results);
+            env->CallStaticVoidMethod(shared_realm_class, partial_sync_cb, nullptr, reinterpret_cast<jlong>(wrapper),
```

Comment:
Don't we risk that these weak references have been GC'ed when the callback is invoked?

Reply:
I can add a check in `SharedRealm#runPartialSyncRegistrationCallback` to see if the weak references returned from JNI are not null (i.e not GC'ed)

```Java
if (callback != null && realm != null) {
            if (error != null) {
               // notify error
                callback.onError(new RealmException(error));
            } else {
               // notify success  
                ...
                callback.onSuccess(results);
            }    
        }
```

@beeender is it safe to call `j_callback_weak.global_ref(env).get()` if the `JavaGlobalWeakRef` has been GC'ed ? I suppose it will return `nullptr` to Java? is this assumption correct?

Reply:
It looks quite complex and breaking our architectures layers since now the `SharedRealm` needs to know the information about `Class<?>`/`RealmResults` which are supposed to live in the `io.realm` package. And also the reference relationship becomes quite complex. 

My idea would be have: `SharedRealm.PartialSyncCallback` then create new instance in `Realm.java`.

I will create a PR to this one to fix this.

Reply:
address this in #5359